### ISSUE TITLE
%column and %char count wrong with states

### ISSUE THREAD

**Issue by vladdu** [jflex-de/jflex#193]
I notice that when using %column and %state with states (for example strings), the values are not updated properly. However, later tokens are correct, so I suppose it's a matter of the update occurring after the actions are executed.`

```
%%
%class Lexer
%unicode
%line
%column
%char
%{
StringBuilder string = new StringBuilder();
private Yytoken sym(MyToken token, String text) {
System.out.println(""+token+" "+yyline+" "+yycolumn+" "+yychar+" <"+text);
    return new Yytoken(token, text, yyline, yycolumn, yychar);
}
%}
StringChar = [^\"\\] | \\.
%state STRING
%%

<YYINITIAL> {
\n { return sym(MyToken.SEP); }
";" { return sym(MyToken.SEP); }
\" { yybegin(STRING); string.setLength(0); string.append("\""); }
[ \t]+ { /* ignore white */ }
[^\n#\"{} \t]+ { return sym(MyToken.WORD, yytext()); }
}

<STRING> {
\" { yybegin(YYINITIAL); return sym(MyToken.WORD, string.append("\"").toString());  }
{StringChar}+  { string.append(yytext()); }
}
```

given input 
```
hello "hello" hello
````

I get output 

````
WORD 0 0 0 <hello
WORD 0 12 12 <"hello"
WORD 0 14 14 <hello
SEP 0 19 19
````

**Comment by schedulix** [comment]
12 is the (zero-based) position of the closing double quote. The code incrementing the column probably isn't aware of what you define to be a token. It only knows if some character has been consumed or not.
Now if some unexpected character occurs, e.g. a '\' isn't handled within a quoted WORD, the error message can use the yycolumn to render a pointer to the offending character instead of the start of the next token.

Like:
```
hello "hello\
------------^
Didn't expect a '\' here
```

instead of (your expectation, if I understand you correctly)
```
hello "hello\
------^
Didn't expect a '\' here
```

**Comment by vladdu** [comment]
Ok, so you mean that I should save the position when encountering the first " and use that? If I have multi-line strings even the line number should be handled that way, I suppose. 

Thanks for the explanation! It is a bit unintuitive, but I can live with that.

**Comment by schedulix** [comment]
Well, though it requires a bit of effort on the programmer's side to determine the token boundaries, it provides more information. Basically, you'll have to save yycolumn and yyline just before a return statement or just before you discard some characters (typically white space, or comments in some programming language). That then would be the starting position of the _next_ token. Both save-variables are initialized with zero, since that's where everything starts.

Then regardless of the number of regexp required to match a single token, you always know where it started. And as I pointed out previously, the actual position within the stream is also known, which is good in case of syntax errors. (start = yycolumn; end = start + yytext.length()).

In short: yycolumn points to the position where the input that matches the current rule started.
I guess this is true for yyline as well, but I'm not sure. Shouldn't be too hard to test.

If you think the issue is resolved, you can close it. But of course you can also wait a little. Maybe some of the Gurus can add some deeper insight into the matter. (I'm just a plain user ;)

**Comment by vladdu** [comment]
Thanks again for the detailed explanation. My understanding is then that the expression "from the beginning of input to the beginning of the current token" found in the docs for yyline, yycolumn and yychar refers to the current regexp being matched, regardless of %state. This is because in the general case, the %state is not visible in the resulting token stream - we can choose to merge together all consecutive tokens inside a %state (like when parsing a string), but that may not always be the case.
